### PR TITLE
nvidia-x11: Package nvidia-bug-report.sh

### LINKED ISSUE #167180: NVIDIA Proprietary driver missing nvidia-bug-report.sh script
### Describe the bug
I want to report an NVIDIA driver bug to nvidia's forums and they require users to submit a log file generated by a script called `nvidia-bug-report.sh` included in their proprietary driver package, however the nix derviation for the drivers doesn't seem to include this script.

### Steps To Reproduce
1. Enable nvidia proprietary driver
2. Try running `nvidia-bug-report.sh`

### Expected behavior
A log file is created

### Notify maintainers
@jonringer 

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.16.18, NixOS, 22.05 (Quokka), 22.05.20220403.1ddbc47`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4`
 - channels(root): `""`
 - channels(tristan): `""`
 - nixpkgs: `/nix/store/8qg31y2v1pn81305kmcqadna7rp6g22w-6lqy3gmi8ajsmljbhsvgx4gbk55152zk-source`
```

### PR DESCRIPTION
**Pull request by robgssp**
###### Description of changes

Install nvidia-bug-report.sh, with a path fix.

Fixes https://github.com/NixOS/nixpkgs/issues/167180

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/nvidia-x11/builder.sh [hunk 1]
```diff
@@ -191,8 +191,8 @@ installPhase() {
                     --set-rpath $out/lib:$libPath $bin/bin/$i
             fi
         done
-        # FIXME: needs PATH and other fixes
-        # install -Dm755 nvidia-bug-report.sh $bin/bin/nvidia-bug-report.sh
+        sed -E 's#/bin/grep#grep#' nvidia-bug-report.sh > nvidia-bug-report.sh.fixed
+        install -Dm755 nvidia-bug-report.sh.fixed $bin/bin/nvidia-bug-report.sh
```
Comment (NickCao) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:195]:
```suggestion
        substituteInPlace nvidia-bug-report.sh --replace /bin/grep ${grep}/bin/grep
        install -Dm755 nvidia-bug-report.sh $bin/bin/nvidia-bug-report.sh
```

Comment (robgssp) [reply pkgs/os-specific/linux/nvidia-x11/builder.sh]: Switched to use substituteInPlace. I recommend keeping the grep invocation path-relative seeing as the rest of the grep calls in the script are relative. The alternative is declaring gnugrep as an explicit dependency in the derivation, which seems unnecessary, especially as there's plenty of other unqualified calls to posix tools in the script.

File: pkgs/os-specific/linux/nvidia-x11/builder.sh [hunk 2]
```diff
@@ -191,8 +191,8 @@ installPhase() {
                     --set-rpath $out/lib:$libPath $bin/bin/$i
             fi
         done
-        # FIXME: needs PATH and other fixes
-        # install -Dm755 nvidia-bug-report.sh $bin/bin/nvidia-bug-report.sh
+        substituteInPlace nvidia-bug-report.sh --replace /bin/grep grep --replace /bin/ls ls
```
Comment (robgssp) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:194]: There was a similar problem with one of the /bin/ls invocations, masked by the script's error handling. Fixed.

Comment (SuperSandro2000) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:194]:
```suggestion
        substituteInPlace nvidia-bug-report.sh \
          --replace /bin/grep grep \
          --replace /bin/ls ls
```

**Comment by thiagokokada** [comment]
Maybe I am doing something wrong, but I am not findinf the `nvidia-bug-report.sh` in the resulting derivation:

```
$ nix-build -A linuxPackages.nvidia_x11
$ ls ./result/bin
nvidia-sleep.sh
```

**Comment by SuperSandro2000** [comment]
Try ``nix-build -A linuxPackages.nvidia_x11.bin``

---

Also this horrible script should be converted to a nix file instead of sourcing stdenv.

**Comment by thiagokokada** [comment]
> nix-build -A linuxPackages.nvidia_x11.bin

Same thing.

File: pkgs/os-specific/linux/nvidia-x11/builder.sh (hunk 1 above)
Comment (nrdxp) [reply pkgs/os-specific/linux/nvidia-x11/builder.sh]: I agree that fixing all the relative calls could be a pain, but perhaps we could use `wrapProgram` and append the expected tools to `PATH` to ensure the dependencies are actually available at runtime?

File: pkgs/os-specific/linux/nvidia-x11/builder.sh [hunk 3]
```diff
@@ -191,8 +191,10 @@ installPhase() {
                     --set-rpath $out/lib:$libPath $bin/bin/$i
             fi
         done
-        # FIXME: needs PATH and other fixes
-        # install -Dm755 nvidia-bug-report.sh $bin/bin/nvidia-bug-report.sh
+        substituteInPlace nvidia-bug-report.sh \
+          --replace /bin/grep grep \
+          --replace /bin/ls ls
+        install -Dm755 nvidia-bug-report.sh $bin/bin/nvidia-bug-report.sh
```
Comment (nrdxp) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:197]: could we maybe add a call to `wrapProgram` here to ensure these (and other tools) are actually available at runtime?

**Review by nrdxp — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/nvidia-x11/builder.sh (hunk 1 above)
Comment (thiagokokada) [reply pkgs/os-specific/linux/nvidia-x11/builder.sh]: We also can use `resholve.mkDerivation` for those scripts. It is kinda of a pain to get it right, but it has the advantage of ensuring that no undeclared dependencies is used.

**Comment by meditans** [comment]
Why is this PR stalled? Is there something wrong with the approach? @nrdxp @NixOS/cuda-maintainers

File: pkgs/os-specific/linux/nvidia-x11/builder.sh
Comment (SomeoneSerge) [inline pkgs/os-specific/linux/nvidia-x11/builder.sh:1]: - @nrdxp's `wrapProgram` suggestion is a reasonable point,
- @meditans do you think you could test the script on a NixOS system prior to merging and report here?

Thanks

Comment (SomeoneSerge) [reply pkgs/os-specific/linux/nvidia-x11/builder.sh:1]: I guess this is not getting done. Maybe we could merge the potentially broken script, which would take us one step closer to somebody reporting how exactly it's broken

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
@ofborg eval

File: pkgs/os-specific/linux/nvidia-x11/builder.sh
Comment (thiagokokada) [reply pkgs/os-specific/linux/nvidia-x11/builder.sh:1]: Merged, let's see if other people can help to fix the script once it is in `master`.

**Pull request merged**